snitun/server/run.py

```diff
@@ -106,9 +117,145 @@ async def _handler(
+        worker_size: Optional[int] = None,
+        throttling: Optional[int] = None,
+    ):
+        """Initialize SniTun Server."""
+        super().__init__()
+
+        self._host: Optional[str] = host
+        self._port: int = port or 443
+        self._fernet_keys: List[str] = fernet_keys
+        self._throttling: Optional[int] = throttling
+        self._worker_size: int = worker_size or (cpu_count() * 2)
+        self._workers: List[ServerWorker] = []
+        self._running: bool = False
+
+        # TCP server
+        self._server: Optional[socket.socket] = None
+        self._poller: Optional[select.epoll] = None
+
+    def start(self) -> None:
+        """Run server."""
+        # Init first all worker, we don't want the epoll on the childs
+        for _ in range(self._worker_size):
+            worker = ServerWorker(self._fernet_keys, throttling=self._throttling)
+            worker.start()
+            self._workers.append(worker)
+
+        self._server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._server.bind((self._host, self._port))
+        self._server.setblocking(False)
+        self._server.listen(80 * 1000)
+
+        self._running = True
+        self._poller = select.epoll()
+        self._poller.register(self._server.fileno(), select.EPOLLIN)
+
+        super().start()
+
+    def stop(self) -> None:
+        """Stop server."""
+        self._running = False
+        self.join()
+
+        # Shutdown all workers
+        for worker in self._workers:
+            worker.shutdown()
+            worker.close()
+
+        self._workers.clear()
+        self._server.close()
+        self._poller.close()
+
+    def run(self) -> None:
+        """Handle incoming connection."""
+        fd_server = self._server.fileno()
+        connections: Dict[int, socket.socket] = {}
+        worker_lb = cycle(self._workers)
+        stale: Dict[int, int] = {}
+
+        while self._running:
+            events = self._poller.poll(1)
+            for fileno, event in events:
+                # New Connection
+                if fileno == fd_server:
+                    con, _ = self._server.accept()
+                    con.setblocking(False)
+
+                    self._poller.register(
+                        con.fileno(), select.EPOLLIN | select.EPOLLHUP | select.EPOLLERR
+                    )
+                    connections[con.fileno()] = con
+                    stale[con.fileno()] = 0
+
+                # Read hello & forward to worker
+                elif event & select.EPOLLIN:
+                    self._poller.unregister(fileno)
+                    con = connections.pop(fileno)
+                    self._process(con, worker_lb)
+
+                # Close
+                else:
+                    self._poller.unregister(fileno)
+                    con = connections.pop(fileno)
+                    con.close()
+
+            # cleanup stale connection
+            for con in tuple(stale.keys()):
```

Comment:
Keys is the default iterator.

```suggestion
            for con in tuple(stale):
```

Any reason to consume the iterator to a tuple? Nothing else is writing to it right?